src/jwks-fetch.js

```diff
@@ -1,54 +1,67 @@
 
-  const cached = cache ? cache.get(cacheKey) : null
+  async function getSecret (signatures) {
+    const { domain, alg, kid } = signatures
+    const cacheKey = `${alg}:${kid}:${domain}`
+    const cachedSecret = cache.get(cacheKey)
 
-  if (cached) {
-    return cached
-  } else if (cache && cached === null) {
-    // null is returned when a previous attempt resulted in the key missing in the JWKs - Do not attemp to fetch again
-    throw new Error(MISSING_KEY_ERROR)
-  }
+    if (cachedSecret) {
+      return cachedSecret
+    } else if (cachedSecret === null) {
+      // null is returned when a previous attempt resulted in the key missing in the JWKs - Do not attemp to fetch again
+      throw new Error(MISSING_KEY_ERROR)
+    }
 
-  // Hit the well-known URL in order to get the key
-  const response = await fetch(`${domain}.well-known/jwks.json`, { timeout: 5000 })
-  const body = await response.json()
+    // Hit the well-known URL in order to get the key
+    const response = await fetch(`${domain}.well-known/jwks.json`, { timeout: 5000 })
+    const body = await response.json()
 
-  if (!response.ok) {
-    const error = new Error(response.statusText)
-    error.response = response
-    error.body = body
+    if (!response.ok) {
+      const error = new Error(response.statusText)
+      error.response = response
+      error.body = body
 
-    throw error
-  }
+      throw error
+    }
 
-  if (!body.keys || body.keys.length === 0) {
-    throw new Error(NO_KEYS_ERROR)
-  }
+    if (!body.keys || body.keys.length === 0) {
+      throw new Error('No keys found in the set.')
+    }
 
-  // Find the key with ID and algorithm matching the JWT token header
-  const key = body.keys.find(k => k.alg === alg && k.kid === kid)
+    // Find the key with ID and algorithm matching the JWT token header
+    const key = body.keys.find(k => k.alg === alg && k.kid === kid)
 
-  if (!key) {
-    // Mark the key as missing
-    cache && cache.set(cacheKey, null)
-    throw new Error(MISSING_KEY_ERROR)
-  }
+    if (!key) {
+      // Mark the key as missing
+      cache.set(cacheKey, null)
+      throw new Error(MISSING_KEY_ERROR)
+    }
+
+    // should we be using https://github.com/Brightspace/node-jwk-to-pem ??
 
-  // should we be using https://github.com/Brightspace/node-jwk-to-pem ??
+    // certToPEM extracted from https://github.com/auth0/node-jwks-rsa/blob/master/src/utils.js
+    const secret = `-----BEGIN CERTIFICATE-----\n${key.x5c[0]}\n-----END CERTIFICATE-----\n`
 
-  // certToPEM extracted from https://github.com/auth0/node-jwks-rsa/blob/master/src/utils.js
-  const secret = `-----BEGIN CERTIFICATE-----\n${key.x5c[0]}\n-----END CERTIFICATE-----\n`
+    // Save the key in the cache
+    cache.set(cacheKey, secret)
+    return secret
+  }
 
-  // Save the key in the cache
-  cache && cache.set(cacheKey, secret)
-  return secret
+  return {
+    getSecret: getSecret,
+    clearCache: () => cache.clear(),
+    cache
```

Comment:
if you're returning an object containing the cache, why return a method to clear the cache?

Reply:
My thinking was to discourage the user to `get()` or `set()` the cache manually and keep the API & documentation simple.  The proposal was to just expose a way to clear the cache to the user, the whole cache object is only used in testing.

Reply:
okay, not a big deal